No deps

greenjustin opened the pull request:
CL series designed to eliminate or mark optional many of the cros-codecs dependencies. We want to make the codec parsing stack as portable as possible in preparation for exposing bindings usable by Chrome.

greenjustin committed:
Add MD5 sums for test data

Add MD5 sums for test data. The example decoder, ccdec, only computes
MD5 sums instead of CRCs, so add some goldens to match.

greenjustin committed:
Add simple MD5 sum comparison script

Simple testing script for comparing ccdec MD5 sum outputs with golden
files. This is not an especially robust testing solution, but it will
help us with development while we get a proper CI set up.

greenjustin committed:
Move NaluReader into utils

Move NaluReader into utils in preparation for use as a generic bitstream
parsing class. Also add the ability to disable emulation prevention byte
skipping.

greenjustin committed:
Switch VP8 parser to use NaluReader

Switch VP8 parser to use NaluReader to break its dependency on BitReader
and Bytes.

greenjustin committed:
Switch VP9 parser to use NaluReader

Switch VP9 parser to use NaluReader to break its dependency on BitReader
and Bytes.

greenjustin committed:
Switch AV1 parser to use NaluReader

Switch AV1 parser to use NaluReader to break its dependency on
BitReader.

greenjustin committed:
Eliminate references to chunk() in NALU code

Eliminate references to cursor.chunk() in NALU code to break the
dependency on the "bytes" crate. This functionality is readily available
in the standard library.

greenjustin committed:
Remove bytes and bitreader dependency

Eliminate dependencies on bytes from utils.rs and remove bytes and
bitreader dependencies from Cargo.toml altogether.

greenjustin committed:
Use GBM allocated frames in tests

greenjustin committed:
Remove anyhow from parsers

greenjustin committed:
Eliminate thiserror from parsers

greenjustin committed:
Eliminate enumn dependency

greenjustin committed:
Separate image processing functions into their own file

greenjustin committed:
Sever parsing code from rest of cros-codecs

Also mark almost all dependencies as optional and disable their
compilation if we just want the parser stack.

greenjustin committed:
Disable all VA-API feature by default.

We want to be able to build the "codecs" directory without any backends.

By default we will only build the parsing stack.

stevecho321 commented on Cargo.toml:
Can we also mention it is disabling "vaapi" in the title?

Figuring out where & how to review here. :)

Meant for "Disable all features by default" commit.

greenjustin replied:
I think github is not very conducive to code reviews. Thankfully we'll be switching to Gerrit soon enough :)

I think it's misleading to say this commit disables VA-API, because it will still build fine when enabling the VA-API feature. This commit just makes it so VA-API is no longer the default build mode.

stevecho321 replied:
Yes, mentioning that will be more clear: VA-API is no longer the default build mode.

stevecho321 commented on src/codec/av1/test_data/av1-annexb.ivf.av1.md5:
empty file? intended?

greenjustin replied:
Deleted in 749aa075d14c98fe475d99383ed6d87044125474
I think this was a broken vector.

stevecho321 commented on src/codec/av1/test_data/test-25fps.ivf.av1.md5:
I see only av1 files have ".ivf" post-fixes.

stevecho321 commented on src/codec/av1/test_data/gen_crcs.sh:
It would be nice to have 1 file for all codecs. But this might be outside of this commit's boundary.

greenjustin replied:
I agree, and I also think this is out of scope for this patch series. I'm planning to overhaul the testing infrastructure entirely once this repo is in a more stable place.

stevecho321 commented on simple_test.py:
```diff
@@ -0,0 +1,40 @@
+    print('Error running ccdec for bitstream ' + bitstream_path)
+    print(ccdec_process.stderr.decode('utf-8'))
+    return
+  test_output = ccdec_process.stdout.decode('utf-8')
```
I don't know what's coding style recommendations for spacing between lines for Rust. 

I would personally prefer to have more spaces between lines for better readability.

Say here between L14 and L15. In general, throughout the code.

greenjustin replied:
This is python, not rust, so the coding style recommendations would be different.

I added some line breaks in 5a1719948912bbba4247f48ff6edaa5f56fa4dc5

stevecho321 replied:
Yes, this is Python! My bad. No wonder the code look more familiar here. :) Thanks.

stevecho321 replied:
I don't have good eye to catch Rust specific syntax issues say for some changes like 
- Remove anyhow from parsers
- Eliminate thiserror from parsers

stevecho321 commented:
> CL series designed to eliminate or mark optional many of the cros-codecs dependencies. We want to make the codec parsing stack as portable as possible in preparation for exposing bindings usable by Chrome.

I assume eventual goal was to move outside of cros-codecs. Are we trying to make parsers "codec" directory to not have any dependency in this code? Little confused about intermediate goal in this CL.

greenjustin committed:
Added some line breaks in simple_test.py

greenjustin committed:
Remove broken test vector av1-annexb.ivf.av1

greenjustin commented:
FYI because of the logistic difficulties with github, I'm going to add additional commits to the merge request addressing comments rather than editing previous commits.

greenjustin commented:
> > CL series designed to eliminate or mark optional many of the cros-codecs dependencies. We want to make the codec parsing stack as portable as possible in preparation for exposing bindings usable by Chrome.
> 
> I assume eventual goal was to move outside of cros-codecs. Are we trying to make parsers "codec" directory to not have any dependency in this code? Little confused about intermediate goal in this CL.

The goal is not to move the "codec" directory outside of cros-codecs, but rather eliminate all dependencies for the codec directory except for the "log" crate. This is because we would like to re-use the codec directory in Chrome in the near future, and all crate dependencies must go through the third party library approval process in order to be added to Chrome, and several of these dependencies pose portability problems on non-ChromeOS platforms.

v4l2r and cros-libva have dependencies that pose obvious portability problems, so part of this effort also involved isolating the "codec" directory in such a way that it can be built with no backends selected.

stevecho321 commented:
> > > CL series designed to eliminate or mark optional many of the cros-codecs dependencies. We want to make the codec parsing stack as portable as possible in preparation for exposing bindings usable by Chrome.
> > 
> > 
> > I assume eventual goal was to move outside of cros-codecs. Are we trying to make parsers "codec" directory to not have any dependency in this code? Little confused about intermediate goal in this CL.
> 
> The goal is not to move the "codec" directory outside of cros-codecs, but rather eliminate all dependencies for the codec directory except for the "log" crate. This is because we would like to re-use the codec directory in Chrome in the near future, and all crate dependencies must go through the third party library approval process in order to be added to Chrome, and several of these dependencies pose portability problems on non-ChromeOS platforms.
> 
> v4l2r and cros-libva have dependencies that pose obvious portability problems, so part of this effort also involved isolating the "codec" directory in such a way that it can be built with no backends selected.

Yes, "codec" directory shouldn't have dependency with v4l2r in theory. I don't recall seeing any dependency. The dependency was mostly with "device" directory.

greenjustin commented:
> > > > CL series designed to eliminate or mark optional many of the cros-codecs dependencies. We want to make the codec parsing stack as portable as possible in preparation for exposing bindings usable by Chrome.
> > > 
> > > 
> > > I assume eventual goal was to move outside of cros-codecs. Are we trying to make parsers "codec" directory to not have any dependency in this code? Little confused about intermediate goal in this CL.
> > 
> > 
> > The goal is not to move the "codec" directory outside of cros-codecs, but rather eliminate all dependencies for the codec directory except for the "log" crate. This is because we would like to re-use the codec directory in Chrome in the near future, and all crate dependencies must go through the third party library approval process in order to be added to Chrome, and several of these dependencies pose portability problems on non-ChromeOS platforms.
> > v4l2r and cros-libva have dependencies that pose obvious portability problems, so part of this effort also involved isolating the "codec" directory in such a way that it can be built with no backends selected.
> 
> Yes, "codec" directory shouldn't have dependency with v4l2r in theory. I don't recall seeing any dependency. The dependency was mostly with "device" directory.

The lib.rs file previously exposed the backend, decoder, and encoder modules, which depended on the v4l2r, cros-libva, and nix crates. Logically there should be no dependency, this refactoring ensures that the build system understands that now.

stevecho321 commented on src/codec/h264/nalu_reader.rs:
not sure where I can comment on commit description. commenting here instead. 

It seems it's intended to make nalu_reader common for all codecs. 

I think NALU is a something we only have with H.264 and HEVC. Isn't it?

You might have intended say more general bitstream parser here, then some names might need to be updated accordingly.

nathan-google replied:
I agree with Steve. The usage of NaluReader I think will lead to confusion. If you want something that reads bits from a slice, then call it BitReader or something similar.

A NaluReader should handle NALU type parsing and boundary discovery and payload slice retrieval. It could be composed of a BitReader, or they both be traits on the same underlying data structure.

stevecho321 replied:
Justin, btw, I noticed there are some changes from Alex & Hiro regarding NaluReader and H.264 & AV1 parser 2 weeks ago. Just in case you missed. Mentioning it here as you are also heavily changing these files. 

https://github.com/greenjustin/cros-codecs/compare/main...chromeos%3Acros-codecs%3Amain

greenjustin replied:
Yes, I noticed those changes. I assume they will conflict heavily with my changes, so I will rebase them and resolve conflicts once we merge these changes.

stevecho321 commented on src/image_processing.rs:
We have been conventionally calling this "image processing" functionalities. 

If this is an opportunity to find better name, will there be a better name? 

It is "image processing" in a broader sense, but it is more close to format conversion here.

greenjustin replied:
I'm open to suggestions, I was simply keeping with Chrome convention. Perhaps this is out of scope for this CL chain though?

stevecho321 replied:
Yes, as it is fine for this CL chain. Up to you if you want to change.

stevecho321 commented on src/utils.rs:
```diff
@@ -9,16 +9,15 @@
 
 use anyhow::anyhow;
 use anyhow::Context;
-use byteorder::ReadBytesExt;
```
What is motivation for below changes? 
- Remove bytes and bitreader dependency
- enumn

greenjustin replied:
The motivation is just to remove the dependency on the bytes, bitreader, and enumn crates. Every crate that we depend on, including transitive dependencies, must go through Chrome's third party library approval process. Plus, it's a security liability. I'd like to eliminate as many dependencies as possible before integration with Chrome to streamline the process.

nathan-google commented on src/utils.rs:
```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
 use crate::PlaneLayout;
 use crate::Resolution;
 
+/// A bit reader for h264 bitstreams. It properly handles emulation-prevention
```
Nit:

/// A bit reader for h264 and h265 bitstreams...

nathan-google commented on src/codec/vp8/bool_decoder.rs:
```diff
@@ -50,19 +49,18 @@ pub enum BoolDecoderError {
+    pub fn new(data: &'a [u8]) -> Self {
         Self {
-            data: Cursor::new(data),
+            data: NaluReader::new(data, false),
```
I think this is too confusing. It feels like there should be a "bit reader" if you want to read bits from a slice. You aren't using any of the NALU parsing features in NaluReader.

greenjustin replied:
I think that's fair, the NALU parsing features confusingly enough are actually in src/codec/h264/nalu.rs. I'll rename this class BitReader. I was originally avoiding using that term because there's a crate called "bitreader" that was also being used in this code, but I removed that dependency a few commits back.

I'll post another commit with that change, since editing previous commits is non-trivial with this development flow...

greenjustin committed:
Rename NaluReader to BitReader to avoid confusion

greenjustin committed:
Re-add av1-annexb.ivf.av1 test vector

This vector was apparently used in unit tests, even though it contains
no frames.

nathan-google commented on src/utils.rs:
```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
+
+    /// Read up to 31 bits from the stream.
+    pub fn read_bits<U: TryFrom<u32>>(&mut self, num_bits: usize) -> Result<U, ReadBitsError> {
+        if num_bits > 31 {
```
Just curious. Why not allow for up to 32? They would fit into `U`.

greenjustin replied:
It would break |read_se| since it would populate the signed bit of i32

greenjustin replied:
Added a comment describing this limitation.

nathan-google commented on src/utils.rs:
```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
+        false
+    }
+
+    pub fn read_ue<U: TryFrom<u32>>(&mut self) -> anyhow::Result<U> {
```
Nit: please add a doc comment.

I.e.

```
/// Reads an Unsigned Exponential golomb coding number from the next bytes in the
/// bitstream. This may advance the state of position within the bitstream even if the
/// read operation is unsuccessful.
```

Or something like that. WDYT?

nathan-google replied:
Similarly for other `read*` or `get*`  methods.

nathan-google commented on src/utils.rs:
```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
+    curr_byte: u8,
+    /// Number of bits remaining in `curr_byte`
+    num_remaining_bits_in_curr_byte: usize,
+    /// Used in epb detection.
```
What is EBP? The annex-B start code prefix or something else?

nathan-google replied:
Ah. Emulation prevention byte. Can you expand the acronym?

nathan-google commented on src/utils.rs:
```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
+        self.data.read_u8().map_err(|_| GetByteError::OutOfBits)
+    }
+
+    fn update_curr_byte(&mut self) -> Result<(), GetByteError> {
```
Optional: I think this private method name could be more self-descriptive.  Maybe "move_to_next_byte"?

nathan-google commented on src/bitstream_utils.rs:
```diff
@@ -0,0 +1,764 @@
+    }
+
+    /// Read a single bit from the stream.
+    pub fn read_bit(&mut self) -> Result<bool, String> {
```
Optional: Maybe the cursor name: `read_bool` is clearer.

nathan-google commented on Cargo.toml:
```diff
@@ -8,22 +8,19 @@ authors = ["The ChromiumOS Authors"]
-vaapi = ["libva"]
-v4l2 = ["v4l2r"]
+default = []
+vaapi = ["libva", "anyhow", "byteorder", "thiserror", "crc32fast", "nix"]
```
Are you planning to remove these dependencies for `vaapi` and `v4l2` in a later change?

greenjustin replied:
Probably not in the near future. These crates are already added to ChromeOS and Android. I'm mostly concerned about dependencies for Chrome, and we're only using the parsers in Chrome for the time being.

nathan-google reviewed (approved):
Overall looks great. I just have a few requests for the BitReader.

greenjustin committed:
Update comments on BitReader

stevecho321 commented:
> Overall looks great. I just have a few requests for the BitReader.

LGTM. Feel free to merge when ready.

greenjustin commented:
Thanks for reviewing everyone!
